### DISCUSSION TITLE
How do I run a set function in v9?

### DISCUSSION THREAD

**Discussion by inspectordanno** [pmndrs/react-spring#1383] [Support] [answered]
I'm experiencing a quite drastic bug after updating from the v9 RC candidate to the current 9.0.0 release, and I think it is related to this issue: I think this question is related to the problems in this issue: https://github.com/pmndrs/react-spring/issues/1359

Unfortunately I cannot post the exact code, but I will attempt to recreate it here.

Before, I would pass a function with an object to `useSpring()`, get a set function out of that, and everything would be fine and dandy. Then, I would use `useEffect()` to run the set function, ie:

```
const config = {
 from: { opacity: 0 },
 to: { opacity: 1 },
 config: { duration: 500 }
}


const [ spring, set ] = useSpring(() => config)

useEffect(() => {
  set(config)
}, [someDependency])

```
Now, this workflow is all broken. The second argument in the destructured array is a ref, not a set function, which is not what the documentation says. Also, I see there is a new dependency array syntax introduced in V9, where you put the array of dependencies in the `useSpring()` call. But after I did this, I was still not getting a set function back - it was still the ref:

`const [ spring, set] = useSpring(config, [someDependency])`

And also, was confused from the v9 documentation: why do I still even need to do`useEffect()` again if the dependency is in `useSpring()`:

```
useEffect(() => {
  set(config)
}, [someDependency])
```

I'm very confused and would appreciate any help. Thank you so much!

**Comment by joshuaellis** [comment, marked as answer]
Would this work?

```js

const [ spring, api ] = useSpring(() => config)

useEffect(() => {
  api.current[0].set(config)
}, [someDependency])

```

It's quite hard to debug without an example. Tbh I'm not super familiar with using deps in the spring.

**Reply by inspectordanno** [reply]
It still doesn't work; says `cannot read property set of undefined (at Controller.set)`

**Reply by joshuaellis** [reply]
I think it would be helpful if you could take this and build a small sandbox demo someone can look at in detail

**Reply by inspectordanno** [reply]
I just figured it out - tried `.start()` instead of `.set()`:

```
useEffect(() => {
  api.current[0].start(config)
}, [someDependency])
```
